spam/spam_tutorial.py

```diff
@@ -41,151 +41,140 @@
+# The class distribution varies slightly from class to class, but all are approximately class-balanced.
 
 # %%
-# Don't truncate text fields in the display
-import pandas as pd
+from collections import Counter
 
-pd.set_option("display.max_colwidth", 0)
+# For clarity, we define constants to represent the class labels for spam, ham, and abstaining.
+ABSTAIN = -1
+HAM = 0
+SPAM = 1
 
-df_dev.sample(5, random_state=123)
+for split_name, df in [("dev", df_dev), ("valid", df_valid), ("test", df_test)]:
+    counts = Counter(df["label"].values)
+    num_points = sum(counts.values())
+    print(f"{split_name.upper():<6} {counts[SPAM] * 100 / num_points:0.1f}% SPAM")
 
 # %% [markdown]
 # ## 2. Write Labeling Functions (LFs)
 
 # %% [markdown]
-# * What's an LF
-#     * Why are they awesome
-# * Can be many types:
-#     * keyword
-#     * pattern-match
-#     * heuristic
-#     * third-party models
-#     * distant supervision
-#     * crowdworkers (non-expert)
-# * Typically an iterative process
-#     * Look at examples for ideas
-#     * Write an LF
-#     * Check performance on dev set
-#     * Balance accuracy/coverage
+# **Labeling functions (LFs) help users encode domain knowledge and other supervision sources programmatically.**
+#
+# LFs are heuristics that take as input a data point and either assign a label to it (in this case, `HAM` or `SPAM`) or abstain (don't assign any label). Labeling functions can be *noisy*: they don't have perfect accuracy and don't have to label every data point.
+#
+# Because their only requirement is that they map a data point a label (or abstain), they can wrap a wide variety of forms of supervision. Examples include, but are not limited to:
+# * Keyword searches: looking for specific words in a sentence
+# * Pattern matching: looking for specific syntactical patterns
+# * Third-party models: using an pre-trained model (usually a model for a different task than the one at hand)
+# * Distant supervision: using external knowledge base
+# * Crowdworker labels: treating each crowdworker as a black-box function that assigns labels to subsets of the data
+#
+# The process of **developing LFs** is iterative and usually consists of:
+# * Writing a function
+# * Estimating its performance by looking at labeled examples in the training set or dev set
+# * Iterating to improve coverage or accuracy as necessary.
+# Balancing accuracy and coverage for specific labeling functions as well as the overall set of LFs developed is often a trade-off, and depending on the use case, users may tend to prefer one over the other.
+#
+# Once multiple LFs have been created, users can look at data points that have received no labels so far (or many conflicting labels) to get ideas for new LFs to write.
+# Another thing to check is whether there are any classes for which there are very few LFs, so that additional LFs can be written to cover that part of the dataset.
```

Comment:
> Another thing to check is whether there are any classes for which there are very few LFs, so that additional LFs can be written to cover that part of the dataset.

Suggestion:
Furthermore, if there are some classes that have votes form very few LFs, users might iterate by writing additional LFs to cover those parts of the dataset.